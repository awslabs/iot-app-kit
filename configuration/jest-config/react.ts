export default {
  setupFilesAfterEnv: ['jest-extended/all'],
  maxWorkers: '50%',
  moduleFileExtensions: ['js', 'ts', 'jsx', 'tsx'],
  collectCoverage: true,
  testEnvironment: 'jsdom',
  transform: {
    '.*\\.(tsx?|jsx?)$': [
      '@swc/jest',
      {
        jsc: {
          transform: {
            react: {
              runtime: 'automatic',
            },
          },
        },
      },
    ],
  },
  transformIgnorePatterns: [],
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy',
  },
};
