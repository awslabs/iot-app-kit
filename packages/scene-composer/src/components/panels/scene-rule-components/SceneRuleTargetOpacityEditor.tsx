import React, { useCallback, useState } from 'react';
import { Input, Grid } from '@cloudscape-design/components';

interface ISceneRuleTargetOpacityEditorProps {
  targetValue: string;
  onChange: (newValue: string) => void;
}

const SceneRuleTargetOpacityEditor: React.FC<ISceneRuleTargetOpacityEditorProps> = ({
  targetValue,
  onChange,
}: ISceneRuleTargetOpacityEditorProps) => {
  const [val, setVal] = useState(targetValue);

  const onInputChange = useCallback(({ detail, target }) => {
    let { value } = detail || target;
    value = value >= 1 ? 1 : value; // disable user from going out of range
    value = value <= 0 ? 0 : value;
    setVal(value);
  }, []);

  const onBlur = useCallback(() => {
    onChange(val);
  }, [val, onChange]);

  return (
    <Grid gridDefinition={[{ colspan: 9 }, { colspan: 3 }]}>
      <Input
        data-testid='tm-opacity-field'
        type='number'
        step={0.01}
        value={val}
        onChange={onInputChange}
        onBlur={onBlur}
      />
    </Grid>
  );
};

SceneRuleTargetOpacityEditor.displayName = 'SceneRuleTargetOpacityEditor';

export default SceneRuleTargetOpacityEditor;
