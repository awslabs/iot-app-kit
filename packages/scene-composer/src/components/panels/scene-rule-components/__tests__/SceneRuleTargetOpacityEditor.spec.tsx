import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import SceneRuleTargetOpacityEditor from '../SceneRuleTargetOpacityEditor';

jest.mock('@awsui/components-react', () => ({
  Grid: 'Grid',
  Input: 'input',
}));

describe('<SceneRuleTargetOpacityEditor>', () => {
  it('render expected field', () => {
    const onChange = jest.fn();
    const value = '1';

    const { container } = render(<SceneRuleTargetOpacityEditor targetValue={value} onChange={onChange} />);
    expect(container).toMatchSnapshot();
  });

  it('update value on change', async () => {
    const onChange = jest.fn();
    const value = '1';

    const { findByTestId } = render(<SceneRuleTargetOpacityEditor targetValue={value} onChange={onChange} />);

    const input = await findByTestId('tm-opacity-field');

    fireEvent.focusIn(input);
    fireEvent.change(input, { target: { value: '0.5' } });
    fireEvent.focusOut(input);

    expect(onChange).toBeCalledWith('0.5');
  });
});
