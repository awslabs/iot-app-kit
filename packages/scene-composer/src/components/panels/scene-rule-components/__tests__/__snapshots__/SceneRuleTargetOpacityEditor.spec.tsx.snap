// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SceneRuleTargetOpacityEditor> render expected field 1`] = `
<div>
  <grid
    griddefinition="[object Object],[object Object]"
  >
    <input
      data-testid="tm-opacity-field"
      step="0.01"
      type="number"
      value="1"
    />
  </grid>
</div>
`;
